File: Advanced/CPP/problem_24.cpp
/*Binary Tree Maximum Path Sum*/
//Solved by aryangupta02092002

#include<iostream>
using namespace std;
struct node {
    int data;
    struct node *left;
    struct node *right;
};

int pathUtil(node* root, int &ans){
    if(root==NULL){
      return 0;
    }
    int tmp1 = root->data;
    int left1 = pathUtil(root->left, ans);
    int right1 = pathUtil(root->right, ans);
    
    int t = max(max(tmp1, tmp1 + left1 + right1), max(tmp1 + left1, tmp1 + right1));
    if(ans < t){
      ans = t;
    }
    return max(max(tmp1 + left1, tmp1 + right1), tmp1);
}

int maxPathSum(node* root) {
    int ans = root->data;
    pathUtil(root, ans);
    return ans;
}

node *newnode(int data){
    node *element=new node();
    element->data=data;
    element->left=NULL;
    element->right=NULL;

    return element;
}

int main(){
    node *root = newnode(-10);
    root->left = newnode(9);
    root->right = newnode(20);
    root->right->left = newnode(15);
    root->right->right = newnode(7);
    
    cout<<maxPathSum(root);
}

File: Advanced/Python/problem_24.py
class Node:

	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None



def findMaxUtil(root):


	if root is None:
		return 0

	l = findMaxUtil(root.left)
	r = findMaxUtil(root.right)

	max_single = max(max(l, r) + root.data, root.data)

	max_top = max(max_single, l+r + root.data)

	findMaxUtil.res = max(findMaxUtil.res, max_top)

	return max_single




def findMaxSum(root):

	findMaxUtil.res = float("-inf")

	findMaxUtil(root)
	return findMaxUtil.res



if __name__ == '__main__':
	root = Node(10)
	root.left = Node(2)
	root.right = Node(10)
	root.left.left = Node(20)
	root.left.right = Node(1)
	root.right.right = Node(-25)
	root.right.right.left = Node(3)
	root.right.right.right = Node(4)


	print "Max path sum is ", findMaxSum(root)

